From 3f962ed9a4079964c48e321fd928a2719038d881 Mon Sep 17 00:00:00 2001
From: P Praneesh <ppranees@codeaurora.org>
Date: Fri, 28 May 2021 23:53:57 +0530
Subject: [PATCH] ath11k: add ampdu id in 802.11 radiotap header

AMPDU aggregate reference number is generated by
driver internally which is same across each
subframe of an ampdu.

For fetching AMPDU-ID, we need to concatenate
ppdu_id from mpdu_info and tlv_usr from tlv heder.
while parsing monitor TLV data with HAL_RX_MPDU_START
TLV tag, ampdu id is fetched from mpdu_info and
updated to corresponding mac80211 structure during
ath11k_update_radiotap.

Signed-off-by: P Praneesh <ppranees@codeaurora.org>
---
 drivers/net/wireless/ath/ath11k/dp_rx.c  |  6 ++++++
 drivers/net/wireless/ath/ath11k/hal_rx.c | 13 ++++++++++++
 drivers/net/wireless/ath/ath11k/hal_rx.h | 16 ++++++++------
 drivers/net/wireless/ath/ath11k/hw.c     | 36 +++++++++++++++++++++++++-------
 drivers/net/wireless/ath/ath11k/hw.h     |  1 +
 5 files changed, 58 insertions(+), 14 deletions(-)

--- a/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -5893,6 +5893,7 @@ static void ath11k_update_radiotap(struc
 {
 	struct ieee80211_supported_band *sband;
 	u8 *ptr = NULL;
+	u16 ampdu_id = ppduinfo->ampdu_id[ppduinfo->userid];
 
 	rxs->flag |= RX_FLAG_MACTIME_START;
 	rxs->signal = ppduinfo->rssi_comb + ATH11K_DEFAULT_NOISE_FLOOR;
@@ -5900,6 +5901,11 @@ static void ath11k_update_radiotap(struc
 	if (ppduinfo->nss)
 		rxs->nss = ppduinfo->nss;
 
+	if (ampdu_id) {
+		rxs->flag |= RX_FLAG_AMPDU_DETAILS;
+		rxs->ampdu_reference = ampdu_id;
+	}
+
 	if (ppduinfo->he_mu_flags) {
 		rxs->flag |= RX_FLAG_RADIOTAP_HE_MU;
 		rxs->encoding = RX_ENC_HE;
--- a/drivers/net/wireless/ath/ath11k/hal_rx.c
+++ b/drivers/net/wireless/ath/ath11k/hal_rx.c
@@ -890,6 +890,13 @@ static u16 ath11k_hal_rx_mpduinfo_get_pe
 	return ab->hw_params.hw_ops->mpdu_info_get_peerid(mpdu_info);
 }
 
+static
+u16 ath11k_hal_rxdesc_get_hal_mpdu_ppdu_id(struct ath11k_base *ab,
+					   struct hal_rx_mpdu_info *mpdu_info)
+{
+	return ab->hw_params.hw_ops->rx_desc_get_hal_ppdu_id(mpdu_info);
+}
+
 static enum hal_rx_mon_status
 ath11k_hal_rx_parse_mon_status_tlv(struct ath11k_base *ab,
 				   struct hal_rx_mon_ppdu_info *ppdu_info,
@@ -1563,6 +1570,12 @@ ath11k_hal_rx_parse_mon_status_tlv(struc
 
 		ppdu_info->mpdu_len += ab->hw_params.hw_ops->rx_desc_get_hal_mpdu_len(mpdu_info);
 
+		if (userid < HAL_MAX_UL_MU_USERS) {
+			ppdu_info->userid = userid;
+			ppdu_info->ampdu_id[userid] =
+				ath11k_hal_rxdesc_get_hal_mpdu_ppdu_id(ab, mpdu_info);
+		}
+
 		break;
 	}
 	case HAL_RXPCU_PPDU_END_INFO: {
--- a/drivers/net/wireless/ath/ath11k/hal_rx.h
+++ b/drivers/net/wireless/ath/ath11k/hal_rx.h
@@ -222,6 +222,8 @@ struct hal_rx_mon_ppdu_info {
 	u32 num_users;
 	u32 mpdu_fcs_ok_bitmap[HAL_RX_NUM_WORDS_PER_PPDU_BITMAP];
 	struct hal_rx_user_status userstats[HAL_MAX_UL_MU_USERS];
+	u8 userid;
+	u16 ampdu_id[HAL_MAX_UL_MU_USERS];
 };
 
 #define HAL_RX_UL_OFDMA_USER_INFO_V0_W0_VALID			BIT(30)
@@ -464,10 +466,11 @@ struct hal_rx_mpdu_info_ipq8074 {
 } __packed;
 
 struct hal_rx_mpdu_info_qcn9074 {
-	__le32 rsvd0[10];
+	__le32 rsvd0[9];
 	__le32 info0;
-	__le32 rsvd1[2];
 	__le32 info1;
+	__le32 rsvd1[2];
+	__le32 info2;
 	__le32 rsvd2[9];
 } __packed;
 
--- a/drivers/net/wireless/ath/ath11k/hw.h
+++ b/drivers/net/wireless/ath/ath11k/hw.h
@@ -330,6 +330,7 @@ struct ath11k_hw_ops {
 	void (*fill_cfr_hdr_info)(struct ath11k *ar,
 				  struct ath11k_csi_cfr_header *header,
 				  struct ath11k_cfr_peer_tx_param *params);
+	u16 (*rx_desc_get_hal_ppdu_id) (struct hal_rx_mpdu_info *mpdu_info);
 };
 
 extern const struct ath11k_hw_ops ipq8074_ops;
--- a/drivers/net/wireless/ath/ath11k/hw.c
+++ b/drivers/net/wireless/ath/ath11k/hw.c
@@ -13,6 +13,7 @@
 #include "hif.h"
 #include "hal.h"
 #include "hw.h"
+#include "hal_rx.h"
 
 /* Map from pdev index to hw mac index */
 static u8 ath11k_hw_ipq8074_mac_from_pdev_id(int pdev_idx)
@@ -990,12 +991,31 @@ static void ath11k_hw_ipq5018_reo_setup(
 }
 
 static u16
+ath11k_hw_ipq8074_rx_desc_get_hal_mpdu_ppdu_id(u8 *tlv_data)
+{
+	struct hal_rx_mpdu_info *mpdu_info =
+		(struct hal_rx_mpdu_info *)tlv_data;
+
+	return FIELD_GET(HAL_RX_MPDU_INFO_INFO0_PPDU_ID,
+			 __le32_to_cpu(u.ipq8074.info0));
+}
+
+static
+u16 ath11k_hw_qcn9074_rx_desc_get_hal_mpdu_ppdu_id(u8 *tlv_data)
+{
+	struct hal_rx_mpdu_info_qcn9074 *mpdu_info =
+		(struct hal_rx_mpdu_info_ipq9074 *)tlv_data;
+
+	return FIELD_GET(HAL_RX_MPDU_INFO_INFO0_PPDU_ID,
+			 __le32_to_cpu(mpdu_info->info0));
+}
+
+static u16
 ath11k_hw_ipq8074_mpdu_info_get_peerid(struct hal_rx_mpdu_info *mpdu_info)
 {
 	u16 peer_id = 0;
-
 	peer_id = FIELD_GET(HAL_RX_MPDU_INFO_INFO0_PEERID,
-			    __le32_to_cpu(mpdu_info->u.ipq8074.info0));
+			    __le32_to_cpu(mpdu_info->u.ipq8074.info1));
 
 	return peer_id;
 }
@@ -1263,6 +1283,7 @@ const struct ath11k_hw_ops ipq8074_ops =
 	.rx_desc_mpdu_start_addr2 = ath11k_hw_ipq8074_rx_desc_mpdu_start_addr2,
 	.get_ring_selector = ath11k_hw_ipq8074_get_tcl_ring_selector,
 	.rx_desc_get_hal_mpdu_len = ath11k_hw_ipq8074_rx_desc_get_hal_mpdu_len,
+	.rx_desc_get_hal_ppdu_id = ath11k_hw_ipq8074_rx_desc_get_hal_mpdu_ppdu_id,
 };
 
 const struct ath11k_hw_ops ipq6018_ops = {
@@ -1312,6 +1333,7 @@ const struct ath11k_hw_ops ipq6018_ops =
 	.rx_desc_dot11_hdr_fields_valid = ath11k_hw_ipq8074_rx_desc_dot11_hdr_fields_valid,
 	.rx_desc_get_dot11_hdr = ath11k_hw_ipq8074_rx_desc_get_dot11_hdr,
 	.rx_desc_get_crypto_header = ath11k_hw_ipq8074_rx_desc_get_crypto_hdr,
+	.rx_desc_get_hal_ppdu_id = ath11k_hw_ipq8074_rx_desc_get_hal_mpdu_ppdu_id,
 };
 
 const struct ath11k_hw_ops qca6390_ops = {
@@ -1362,6 +1384,7 @@ const struct ath11k_hw_ops qca6390_ops =
 	.rx_desc_get_dot11_hdr = ath11k_hw_ipq8074_rx_desc_get_dot11_hdr,
 	.rx_desc_get_crypto_header = ath11k_hw_ipq8074_rx_desc_get_crypto_hdr,
 	.rx_desc_get_ip_valid = ath11k_hw_ipq8074_rx_desc_get_ip_valid,
+	.rx_desc_get_hal_ppdu_id = ath11k_hw_ipq8074_rx_desc_get_hal_mpdu_ppdu_id,
 };
 
 const struct ath11k_hw_ops qcn9074_ops = {
@@ -1403,6 +1426,7 @@ const struct ath11k_hw_ops qcn9074_ops =
 	.rx_desc_mpdu_start_addr2 = ath11k_hw_ipq9074_rx_desc_mpdu_start_addr2,
 	.get_ring_selector = ath11k_hw_ipq8074_get_tcl_ring_selector,
 	.rx_desc_get_hal_mpdu_len = ath11k_hw_qcn9074_rx_desc_get_hal_mpdu_len,
+	.rx_desc_get_hal_ppdu_id = ath11k_hw_qcn9074_rx_desc_get_hal_mpdu_ppdu_id,
 };
 
 const struct ath11k_hw_ops wcn6855_ops = {
@@ -1452,6 +1476,7 @@ const struct ath11k_hw_ops wcn6855_ops =
 	.rx_desc_get_dot11_hdr = ath11k_hw_ipq8074_rx_desc_get_dot11_hdr,
 	.rx_desc_get_crypto_header = ath11k_hw_ipq8074_rx_desc_get_crypto_hdr,
 	.fill_cfr_hdr_info = ath11k_hw_ipq8074_fill_cfr_hdr_info,
+	.rx_desc_get_hal_ppdu_id = ath11k_hw_ipq8074_rx_desc_get_hal_mpdu_ppdu_id,
 };
 
 const struct ath11k_hw_ops wcn6750_ops = {
@@ -1494,6 +1519,7 @@ const struct ath11k_hw_ops wcn6750_ops =
 	.rx_desc_mpdu_start_addr2 = ath11k_hw_ipq9074_rx_desc_mpdu_start_addr2,
 	.get_ring_selector = ath11k_hw_wcn6750_get_tcl_ring_selector,
 	.rx_desc_set_msdu_len = ath11k_hw_qcn9074_rx_desc_set_msdu_len,
+	.rx_desc_get_hal_ppdu_id = ath11k_hw_qcn9074_rx_desc_get_hal_mpdu_ppdu_id,
 };
 
 /* IPQ5018 hw ops is similar to QCN9074 except for the dest ring remap */
@@ -1546,6 +1572,7 @@ const struct ath11k_hw_ops ipq5018_ops =
 	.rx_desc_get_dot11_hdr = ath11k_hw_qcn9074_rx_desc_get_dot11_hdr,
 	.rx_desc_get_crypto_header = ath11k_hw_qcn9074_rx_desc_get_crypto_hdr,
 	.fill_cfr_hdr_info = ath11k_hw_qcn9074_fill_cfr_hdr_info,
+	.rx_desc_get_hal_ppdu_id = ath11k_hw_qcn9074_rx_desc_get_hal_mpdu_ppdu_id,
 };
 
 const struct ath11k_hw_ops qcn6122_ops = {
